### DISCUSSION TITLE
Multupart endpoint with dictionary data

### DISCUSSION THREAD

**Discussion by aKenan** [reactiveui/refit#1639] [Q&A] [answered]
I am trying to invoke Multipart POST endpoint using refit that expects dictionary data. 
```
[HttpPost]
[Consumes("multipart/form-data")]
public async Task<IActionResult> CreateAndUploadDocument([FromForm] Request command)
{
    //
}


class Request
{
    public Dictionary<string, string> Metadata { get; set; } = null!;
    public IFormFile File { get; set; }
}

// I try to invoke it like this

[Multipart]
[Post("/api/doc")]
Task<ApiResponse<ApiSuccessfulResponse<Response<Guid>>>> CreateAndUploadAsync(Dictionary<string, string> metadata, [AliasAs("file")] StreamPart fileStream);
```

File is recieved properly, but dictionary is always empty. Any suggestions how I can parse dictionary values?

**Comment by glennawatson** [comment, marked as answer]
This is expected. Refit does not decompose a Dictionary<string, string> into separate multipart form fields. In a [Multipart] method, a non-stream/non-string/non-byte[] parameter falls through to the content serializer, so the whole dictionary gets written as a single JSON part. ASP.NET model binding for [FromForm] Dictionary<string, string> expects discrete fields like Metadata[key1]=v1, Metadata[key2]=v2, so it binds nothing and you see an empty dictionary.

Options:

1. Send each entry as its own string part. The simplest reliable approach is to build the parts yourself. Accept the metadata as individual StringPart/parameters, or pass a Dictionary<string, string> and expand it with AliasAs per key. Refit will add a string parameter as a form field with the parameter/alias name:

```csharp
[Multipart]
[Post("/api/doc")]
Task<...> CreateAndUploadAsync(
    [AliasAs("Metadata[type]")] string metadataType,
    [AliasAs("Metadata[owner]")] string metadataOwner,
    [AliasAs("file")] StreamPart fileStream);
```

Each string becomes a separate form field, and the bracket naming matches ASP.NET's dictionary binding.

2. If keys are dynamic, drop [Multipart] on the Refit side and build a MultipartFormDataContent yourself (one StringContent per kv pair named Metadata[<key>]), passing it as HttpContent.

3. If you control the server, accept the metadata as a single JSON string field and deserialize it server-side; then a single string part works.

So the dictionary needs to be flattened into individual form fields (with [key] naming) for [FromForm] dictionary binding to populate it.
